qualtran/bloqs/qft/approximate_qft.py

```diff
@@ -0,0 +1,93 @@
+from typing import Callable
+
+import attrs
+import cirq
+from numpy.typing import NDArray
+
+from qualtran import GateWithRegisters, QFxp, QUInt, Signature
+from qualtran.bloqs.arithmetic.multiplication import PlusEqualProduct
+
+
+@attrs.frozen
+class ApproximateQFT(GateWithRegisters):
+    r"""
+    An approximate QFT, using the phase gradient trick, where we delete controlled z-power gates
+    smaller than a user-defined threshold.
+
+    Given a b-bit phase gradient state $|\phi\rangle$ prepared as
+
+    $$
+        |\phi\rangle = \frac{1}{\sqrt{2^{b}}} \sum_{k=0}^{2^{b} - 1} \omega_{b}^{-k} |k\rangle
+    $$
+
+    The QFT uses exponentially small z-power gates. In practice, it is often sufficient to perform
+    an approximate qft, where z-power gates smaller than a certain threshold are dropped.
+
+    Phase gradient rotations can be synthesized via additions into the phase gradient register.
+    This leads to significant reductions in T/Toffoli complexity and requires 0 arbitrary
+    rotations (given a one-time cost to prepare the gradient register). See the linked reference
+    for more details.
+
+    Args:
+        bitsize: Size of input register to apply QFT on.
+        b: The size of the phase gradient register. This is a function of the bitsize
+           and defaults to being math.ceil(math.log2(bitsize)).
+        with_reverse: Whether or not to include the swaps at the end
+            of the circuit decomposition that reverse the order of the
+            qubits. If True, the swaps are inserted. Defaults to True.
+            These are technically necessary in order to perform the
+            correct effect, but can almost always be optimized away by just
+            performing later operations on different qubits.
+
+    References:
+        [Turning Gradients into Additions into QFTs](https://algassert.com/post/1620)
+    """
+
+    bitsize: int
+    b: Callable = lambda n: math.ceil(math.log2(n))
+    with_reverse: bool = True
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        phase_grad_bitsize = self.b(self.bitsize)
+        assert (phase_grad_bitsize > 0, "b_func must return a positive value for b")
+        return Signature.build_from_dtypes(
+            q=QUInt(self.bitsize), phase_grad=QFxp(phase_grad_bitsize, phase_grad_bitsize)
+        )
+
+    def decompose_from_registers(
+        self, *, context: cirq.DecompositionContext, **quregs: NDArray[cirq.Qid]
+    ) -> cirq.OP_TREE:
+        if self.bitsize == 1:
+            yield cirq.H(*quregs['q'])
+            return
+        q, phase_grad = quregs['q'], quregs['phase_grad']
+        for i in range(len(q)):
+            if i == 0:
+                yield cirq.H(q[i])
+                continue
+            addition_bitsize = min(i, len(phase_grad) - 1)
+            a, b = q[:addition_bitsize], phase_grad[: addition_bitsize + 1]
+            yield PlusEqualProduct(addition_bitsize, 1, addition_bitsize + 1).on_registers(
```

Comment:
@tanujkhattar originally, I was using AddIntoPhaseGrad using the following line of code: `yield AddIntoPhaseGrad(addition_bitsize, addition_bitsize + 1, right_shift=1).on_registers(x=a[::-1], phase_grad=b).controlled_by(q[i])`

However, the T-count is hard to access from this construction. And after all, a controlled-addition is exactly the same thing as multiplying the summand by the control. This might be the most efficient way to implement a controlled-addition?